### PR TITLE
[flang]Lower by BaseAddressByValueAttribute arguments

### PR DESCRIPTION
**Pull request by Leporacanthicus**
Allow value attribute for function arguments, and make a copy
of the actual argument if that is the case.

Implement testing for integer and array of integer arguments.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -2794,7 +2802,23 @@ class ScalarExprLowering {
         // address without making a copy.
         argAddr = readIfBoxValue(argAddr);
 
-        if (arg.passBy == PassBy::BaseAddress) {
+        if (arg.passBy == PassBy::BaseAddressValueAttribute) {
+          mlir::Value temp;
+          if (isArray(*expr)) {
+            auto val = genBoxArg(*expr);
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2808]: Is the array copy really needed if `actualArgIsVariable` is `false` for arrays ?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 2]
```diff
@@ -2711,7 +2717,9 @@ class ScalarExprLowering {
         continue;
       }
       const bool actualArgIsVariable = Fortran::evaluate::IsVariable(*expr);
-      if (arg.passBy == PassBy::BaseAddress || arg.passBy == PassBy::BoxChar) {
+      if (arg.passBy == PassBy::BaseAddress ||
+          arg.passBy == PassBy::BaseAddressValueAttribute ||
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2721]: I would not stick this case here since all the copy-in/copy-out part is unrelated to this case. I would just place it above. 
```
if (arg.passBy == PassBy::BaseAddressValueAttribute) {
} else if (arg.passBy == PassBy::BaseAddress || arg.passBy == PassBy::BaseAddressValueAttribute) {
} else if ...
```

**Review by jeanPerier — COMMENTED** [review]
Thanks for improving the `VALUE` attribute support here. Your PR makes me realize that lowering is not covering the the assumed shape with VALUE attribute case.
We should have a proper TODO in CallInterface.cpp for this  case [here](https://github.com/flang-compiler/f18-llvm-project/blob/dc8541eaca6c7ea1eea945558b19043bb64fa8a5/flang/lib/Lower/CallInterface.cpp#L839) when `isValueAttr` is `true`.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (Leporacanthicus) [reply flang/lib/Lower/ConvertExpr.cpp:2808]: > Thanks for improving the `VALUE` attribute support here. Your PR makes me realize that lowering is not covering the the assumed shape with VALUE attribute case. We should have a proper TODO in CallInterface.cpp for this case [here](https://github.com/flang-compiler/f18-llvm-project/blob/dc8541eaca6c7ea1eea945558b19043bb64fa8a5/flang/lib/Lower/CallInterface.cpp#L839) when `isValueAttr` is `true`.

I'll take a look at that in a bit - I'm away for a fair few days in the next three weeks, so not sure I can get much done. Adding a TODO may fall within what I can achieve tho'. ;)

I have removed the copy if we already have a copy - at least for the test case I made right now.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 2 above)
Comment (Leporacanthicus) [reply flang/lib/Lower/ConvertExpr.cpp:2721]: Yeah, makes sense. Moved it up from that portion.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -2711,6 +2717,28 @@ class ScalarExprLowering {
...
+          }
+        } else {
+          mlir::Value val = fir::getBase(genval(*expr));
+          temp = builder.createTemporary(
+              loc, val.getType(),
+              llvm::ArrayRef<mlir::NamedAttribute>{
+                  Fortran::lower::getAdaptToByRefAttr(builder)});
+          builder.create<fir::StoreOp>(loc, val, fir::getBase(temp));
+        }
+        caller.placeInput(arg, fir::getBase(temp));
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2739]: No need to call `fir::getBase` on something that is already an mlir::Value (calling it creates an fir::ExtendedValue from the Value that is passed to getBase, and then extract that Value back.)

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -2711,6 +2717,28 @@ class ScalarExprLowering {
...
+            genArrayCopy(copy, val);
+            temp = fir::getBase(copy);
+          }
+        } else {
+          mlir::Value val = fir::getBase(genval(*expr));
+          temp = builder.createTemporary(
+              loc, val.getType(),
+              llvm::ArrayRef<mlir::NamedAttribute>{
+                  Fortran::lower::getAdaptToByRefAttr(builder)});
+          builder.create<fir::StoreOp>(loc, val, fir::getBase(temp));
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2737]: No need for this `fir::getBase()` too.

**Review by jeanPerier — APPROVED** [review]
Two small nits about useless `fir::getBase()` calls, appart from that, this looks good to me, thanks ! No worries for the TODO, I will add it myself.

**Review by jeanPerier — APPROVED** [review]
